src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java

```diff
@@ -0,0 +1,428 @@
+
+        protected abstract UNIT_COLLECTOR makeUnknownCollector();
+
+        public Distributor(final List<SAMReadGroupRecord> rgRecs) {
+            collectors = new LinkedHashMap<>();
+            for(final SAMReadGroupRecord rg : rgRecs) {
+                final String key = getKey(rg);
+                if(!collectors.containsKey(key)) {
+                    collectors.put(key, makeCollector(rg));
+                }
+            }
+        }
+
+        /** Call acceptRecord(args) on the record collector identified by getKey */
+        public void acceptRecord(final ARGTYPE args, final SAMReadGroupRecord rg) {
+
+            String key = UNKNOWN;
+            if(rg != null) {
+                final String computedKey = getKey(rg);
+                if(computedKey != null) {
+                    key = computedKey;
+                }
+            }
+            UNIT_COLLECTOR collector = collectors.get(key);
+            if (collector == null) {
+                if (!UNKNOWN.equals(key)) {
+                    throw new GATKException("Could not find collector for " + key);
+                }
+                collector = makeUnknownCollector();
+                collectors.put(key, collector);
+            }
+            collector.acceptRecord(args);
+        }
+
+        /** Call finish on each PerUnitMetricCollector in this Aggregate Collector */
+        public void finish() {
+            for(final UNIT_COLLECTOR collector : collectors.values()) {
+                collector.finish();
+            }
+        }
+
+        /**
+         * Combine/merge the source distributor into this distributor by merging the collector maps.
+         * @param source source Distributor
+         * @param reMap map entry merge (reMap) function for combining distributor map entries. This
+         *              is usually the combineUnit method of MultiLevelCollectorReducible.
+         */
+        public void combine(Distributor source, BiFunction<UNIT_COLLECTOR, UNIT_COLLECTOR, UNIT_COLLECTOR> reMap) {
+            collectors.keySet().forEach(k -> collectors.merge(k, source.collectors.get(k), reMap));
+        }
+
+        /** Add all records to the MetricsFile passed in, this will happen in the order they were
+         * found in the input ReadGroup records */
+        public void addToFile(final MetricsFile<METRIC_TYPE, HISTOGRAM_KEY> file) {
+            for(final UNIT_COLLECTOR collector : collectors.values()) {
+                collector.addMetricsToFile(file);
+            }
+        }
+    }
+
+    /** A dummy Distributor to handle the ALL_READS accumulation level.  No distribution is required
+     * since there should only ever be one PerUnitMetricCollector for ALL_READS.
+     */
+    private class AllReadsDistributor extends Distributor implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        public AllReadsDistributor(final List<SAMReadGroupRecord> rgRecs) {
```

Comment:
the rgRegs are ignored, should you remove them from the constructor's signature?